server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -289,28 +291,66 @@ static class Compatible extends FiltersAggregator {
 
         @Override
         protected LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCtx, LeafBucketCollector sub) throws IOException {
-            if (QueryToFilterAdapter.MatchesNoDocs(filters()) && otherBucketKey == null) {
+            if (QueryToFilterAdapter.matchesNoDocs(filters()) && otherBucketKey == null) {
                 return LeafBucketCollector.NO_OP_COLLECTOR;
             }
-
-            IntPredicate[] docFilters = new IntPredicate[filters().size()];
+            final DocIdSetIterator[] filteredDocIdSetIterators = new DocIdSetIterator[filters().size()];
             for (int filterOrd = 0; filterOrd < filters().size(); filterOrd++) {
-                docFilters[filterOrd] = filters().get(filterOrd).matchingDocIds(aggCtx.getLeafReaderContext());
+                filteredDocIdSetIterators[filterOrd] = filters().get(filterOrd).matchingDocIdSetIterator(aggCtx.getLeafReaderContext());
             }
+
+            final boolean useCompetitiveIterator = (parent == null && otherBucketKey == null);
+            final int maxDocId = aggCtx.getLeafReaderContext().reader().maxDoc();
+
+            // Tracks the last matched doc per filter.
+            final int[] lastMatches = new int[filteredDocIdSetIterators.length];
+            Arrays.fill(lastMatches, -1);
+
+            /*
+             * Even if the competitive iterator is applicable, it may not be used immediately.
+             * We track the last doc id we've processed to effectively switch between the iterator and the direct use
+             * of the filteredDocIdSetIterators in the leaf collector.
+             */
+            final DocId currentDocID = new DocId();
+            currentDocID.set(-1);
+
             return new LeafBucketCollectorBase(sub, null) {
+
                 @Override
                 public void collect(int doc, long bucket) throws IOException {
                     boolean matched = false;
-                    for (int i = 0; i < docFilters.length; i++) {
-                        if (docFilters[i].test(doc)) {
-                            collectBucket(sub, doc, bucketOrd(bucket, i));
-                            matched = true;
+                    // Check if the competitive iterator has matched the current doc.
+                    if (useCompetitiveIterator && doc == currentDocID.get()) {
```

Comment:
Is this branch worth the effort?

Reply:
The good thing here is that the logic is the same, whether the competitive iterator doesn't apply or it hasn't been used and the collector was invoked directly.

Reply:
Maybe it makes sense to have two have two implementations. One when we know we never can use competitive iterator and one when we know competitive iterator can be used?

Reply:
Done.